File: packages/provide-uterm-go/fileio/fileio.go
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

// Package fileio provides file I/O helpers for secure recording sinks, BBS
// screen files, and color palettes. Port of provide.uterm.file_io.
package fileio

import (
	"encoding/json"
	"fmt"
	"math"
	"os"
	"path/filepath"
	"syscall"

	"github.com/provide-io/provide-uterm/packages/provide-uterm-go/ansi"
)

func ensureOwnerOnlyDir(directory string, mode os.FileMode) error {
	if err := os.MkdirAll(directory, mode); err != nil {
		return err
	}
	// MkdirAll only applies the mode when it creates the directory — a
	// pre-existing world-readable dir would stay that way. Re-tighten
	// unconditionally so recording filenames are never enumerable by other
	// local users.
	return os.Chmod(directory, mode)
}

// SecureOpenAppend creates/opens path for append with owner-only permissions
// and no symlink following:
//
//   - O_NOFOLLOW refuses to open a symlink at the target path, so an attacker
//     who pre-creates one cannot redirect writes; a symlinked recording path
//     is an attack or misconfiguration and fails loudly (ELOOP).
//   - The create mode only applies on creation, so a pre-existing loose file
//     would keep its mode; fchmod on the just-opened fd re-tightens it to
//     0600 with no TOCTOU race (it targets the fd, not the path).
//   - A regular-file check rejects a fifo/device pre-created at the path.
func SecureOpenAppend(path string) (*os.File, error) {
	return SecureOpenAppendMode(path, 0o600, 0o700)
}

// SecureOpenAppendMode is SecureOpenAppend with explicit file and directory
// modes.
func SecureOpenAppendMode(path string, mode, dirMode os.FileMode) (*os.File, error) {
	if err := ensureOwnerOnlyDir(filepath.Dir(path), dirMode); err != nil {
		return nil, err
	}
	f, err := os.OpenFile(path, os.O_WRONLY|os.O_CREATE|os.O_APPEND|syscall.O_NOFOLLOW, mode)
	if err != nil {
		return nil, err
	}
	info, err := statFile(f)
	if err != nil {
		_ = f.Close()
		return nil, err
	}
	if !info.Mode().IsRegular() {
		_ = f.Close()
		return nil, fmt.Errorf("refusing to open non-regular recording sink: %s", path)
	}
	if err := chmodFile(f, mode); err != nil {
		_ = f.Close()
		return nil, err
	}
	return f, nil
}

// statFile and chmodFile are indirection points so tests can exercise the
// fd-level error paths, which cannot be triggered deterministically through
// the filesystem.
var (
	statFile  = func(f *os.File) (os.FileInfo, error) { return f.Stat() }
	chmodFile = func(f *os.File, mode os.FileMode) error { return f.Chmod(mode) }
)

// LoadANS loads a .ans file (BBS ANSI art) as latin-1 text: every byte maps
// one-to-one onto codepoints U+0000-U+00FF.
func LoadANS(path string) (string, error) {
	raw, err := os.ReadFile(path)
	if err != nil {
		return "", err
	}
	runes := make([]rune, len(raw))
	for i, b := range raw {
		runes[i] = rune(b)
	}
	return string(runes), nil
}

// LoadTxt loads a plain UTF-8 text file.
func LoadTxt(path string) (string, error) {
	raw, err := os.ReadFile(path)
	if err != nil {
		return "", err
	}
	return string(raw), nil
}

// LoadPalette loads a JSON 256-color palette (list of 16 ints 0-255). An
// empty path returns a copy of ansi.DefaultPalette.
func LoadPalette(path string) ([]int, error) {
	if path == "" {
		out := make([]int, len(ansi.DefaultPalette))
		copy(out, ansi.DefaultPalette)
		return out, nil
	}
	raw, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	var data []any
	if err := json.Unmarshal(raw, &data); err != nil || len(data) != 16 {
		return nil, fmt.Errorf("palette map must be a JSON list of 16 integers")
	}
	out := make([]int, 0, 16)
	for _, v := range data {
		f, ok := v.(float64)
		if !ok || f != math.Trunc(f) || f < 0 || f > 255 {
			return nil, fmt.Errorf("palette map values must be integers in 0..255")
		}
		out = append(out, int(f))
	}
	return out, nil
}

File: packages/provide-uterm-ts/src/file-io/file-io.ts
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

/**
 * File I/O helpers for loading BBS screen files and colour palettes, and for
 * opening capture sinks safely.
 *
 * Port of the Python module `provide.uterm.file_io` and the Go package
 * `fileio`.
 */

import {
  chmodSync,
  closeSync,
  constants,
  fchmodSync,
  fstatSync,
  mkdirSync,
  openSync,
  readFileSync,
  writeSync,
} from "node:fs";
import { dirname } from "node:path";
import { DEFAULT_PALETTE } from "../ansi/index.ts";

/** Permission options for the secure-open helpers. */
export interface SecureCreateOptions {
  /** File mode. Defaults to owner read/write only. */
  mode?: number;
  /** Parent directory mode. Defaults to owner-only. */
  dirMode?: number;
}

/**
 * Create `directory` and its parents, then re-apply `mode`.
 *
 * `mkdir`'s mode argument only takes effect when it actually creates the
 * directory, so a pre-existing group-readable recordings directory would keep
 * its bits and leave session filenames enumerable by other local users. The
 * explicit `chmod` afterwards enforces the mode either way.
 */
function ensureOwnerOnlyDir(directory: string, mode: number): void {
  mkdirSync(directory, { mode, recursive: true });
  chmodSync(directory, mode);
}

/**
 * Open `path` for append with owner-only permissions, refusing to follow a
 * symlink.
 *
 * Creating the file through `open` with the mode set avoids the window an
 * `open`-then-`chmod` pair leaves, during which the file is briefly readable
 * by others. `O_NOFOLLOW` means a symlink planted at the target path fails
 * the open rather than redirecting the write, and the regular-file check
 * rejects a FIFO or device planted there instead.
 *
 * @returns An open file descriptor. The caller owns closing it.
 */
export function secureCreate(path: string, options: SecureCreateOptions = {}): number {
  const mode = options.mode ?? 0o600;
  const dirMode = options.dirMode ?? 0o700;
  ensureOwnerOnlyDir(dirname(path), dirMode);
  const fd = openSync(path, constants.O_WRONLY | constants.O_CREAT | constants.O_APPEND | constants.O_NOFOLLOW, mode);
  try {
    // Defensive, and carried over from the reference. The kernel already
    // rejects a write-open of a directory with EISDIR, and a FIFO blocks
    // rather than returning, so reaching this needs a device node the test
    // suite cannot create unprivileged. It stays as the backstop it is.
    /* v8 ignore next 3 */
    if (!fstatSync(fd).isFile()) {
      throw new Error(`Refusing to open non-regular recording sink: ${path}`);
    }
    // Re-apply the mode: it only takes effect on creation, so a file that
    // already existed keeps whatever bits it had.
    fchmodSync(fd, mode);
  } catch (error) {
    closeSync(fd);
    throw error;
  }
  return fd;
}

/** An append-only text sink. */
export interface AppendHandle {
  /** Append `text` as UTF-8. */
  writeSync(text: string): void;
  /** Close the underlying descriptor. */
  close(): void;
}

/** Open `path` for UTF-8 append using {@link secureCreate}. */
export function secureOpenAppend(path: string, options: SecureCreateOptions = {}): AppendHandle {
  const fd = secureCreate(path, options);
  return {
    writeSync(text: string): void {
      writeSync(fd, Buffer.from(text, "utf-8"));
    },
    close(): void {
      closeSync(fd);
    },
  };
}

/**
 * Load a `.ans` file (BBS ANSI art).
 *
 * Defaults to latin-1, the standard encoding for BBS ANSI art, which maps
 * every byte to a code point so nothing is lost.
 */
export function loadAns(path: string, encoding: BufferEncoding = "latin1"): string {
  return readFileSync(path).toString(encoding);
}

/** Load a plain text file, UTF-8 by default. */
export function loadTxt(path: string, encoding: BufferEncoding = "utf-8"): string {
  return readFileSync(path).toString(encoding);
}

/**
 * Load a JSON 256-colour palette: a list of exactly 16 integers in 0..255.
 *
 * @param path Palette file, or `null` for a copy of the default palette.
 * @throws {Error} If the file is not a list of exactly 16 in-range integers.
 */
export function loadPalette(path: string | null): number[] {
  if (path === null) {
    return [...DEFAULT_PALETTE];
  }
  const data: unknown = JSON.parse(readFileSync(path, "utf-8"));
  if (!Array.isArray(data) || data.length !== 16) {
    throw new Error("palette map must be a JSON list of 16 integers");
  }
  const out: number[] = [];
  for (const value of data) {
    // A boolean is an int in Python, but the reference's isinstance check
    // admits it; refusing it here is the stricter and safer reading, and no
    // real palette carries one.
    if (typeof value !== "number" || !Number.isInteger(value) || value < 0 || value > 255) {
      throw new Error("palette map values must be integers in 0..255");
    }
    out.push(value);
  }
  return out;
}
